Uneven Mango Ferret

Medium

# Potential Silent Failure in `externalCall` Function Due to Unchecked Call Result

## Summary
The externalCall function performs an external call to a specified address using the provided calldata. However, the issues is external calls using the call method return false rather than reverting when they fail. This can lead to silent failures, leaving the system in an inconsistent state. The function currently emits an event with the success status but does not handle the case where the external call fails, which can mask underlying issues
## Vulnerability Detail
The vulnerability arises from the use of the call low-level function. External calls are risky because they do not revert on failure; instead, they return false. If the external call fails due to gas exhaustion, invalid opcode, or other reasons, the externalCall function will simply emit an event indicating the failure without reverting the transaction. This can lead to issues, especially if the external call is critical to the system's operations.
## Impact
Silent Failures: External calls that fail will not revert the transaction, leading to silent failures. This can mask underlying issues and make debugging difficult.
Inconsistent State: Operations that rely on successful external calls may proceed with incorrect assumptions, resulting in an inconsistent state.
Security Risks: Critical operations may not execute as expected, leading to potential security vulnerabilities.
## Code Snippet
https://github.com/sherlock-audit/2024-06-mellow/blob/main/mellow-lrt/src/Vault.sol#L249

```solidity
   function externalCall(
        address to,
        bytes calldata data
    ) external nonReentrant returns (bool success, bytes memory response) {
        _requireAtLeastOperator();
        if (configurator.isDelegateModuleApproved(to)) revert Forbidden();
        IValidator validator = IValidator(configurator.validator());
        validator.validate(
            msg.sender,
            address(this),
            abi.encodeWithSelector(msg.sig, to, data)
        );
        validator.validate(address(this), to, data);
        (success, response) = to.call(data);
        emit ExternalCall(to, data, success, response);
    }
```

## Tool used

Manual Review

## Recommendation
the `externalCall` function should include a check for the success of the call and revert if it fails. 

```diff
function externalCall(
    address to,
    bytes calldata data
) external nonReentrant returns (bool success, bytes memory response) {
    _requireAtLeastOperator();
    if (configurator.isDelegateModuleApproved(to)) revert Forbidden();
    IValidator validator = IValidator(configurator.validator());
    validator.validate(
        msg.sender,
        address(this),
        abi.encodeWithSelector(msg.sig, to, data)
    );
    validator.validate(address(this), to, data);
    (success, response) = to.call(data);
+    require(success, "External call failed");
    emit ExternalCall(to, data, success, response);
}
```
